### PR TITLE
docs(join): Documentation for the new join package.

### LINKED ISSUE #4631: Document `join` package
DOD:
- Doc comments have been added to each item in the new `join` package, and they pass the `check-fluxdoc` CI job
- Those docs have been reviewed and approved by someone on the docs team

### PR DESCRIPTION
**Pull request by scbrickley**
Adds doc comments for the new join package, introduced in #4690

Closes #4631

### CODE REVIEW

**Review by nathanielc — CHANGES_REQUESTED** [review]
Looks good some comments/questions below

File: stdlib/join/join.flux [hunk 1]
```diff
@@ -1,15 +1,194 @@
-// Package join is under active development and is not yet ready for public consumption.
+// Package join provides functions that join two table streams together.
+//
+// ## Outer Joins
```
Comment (sanderson) [inline stdlib/join/join.flux:3]:
```suggestion
// ## Outer joins
```

File: stdlib/join/join.flux [hunk 2]
```diff
@@ -1,15 +1,194 @@
...
+//   or more equality comparisons between a property of `l` and a property of `r`,
+//   each chained together by the `and` operator.
+// - as: Function that takes a left and a right record (`l` and `r` respectively), and returns a record.
+//   The returned record will be included in the final output.
+// - method: String that specifies the join method.
+//   Must be `"inner"`, `"left"`, `"right"`, or `"full"`
+//
+// ## Examples
+//
+// ### Perform an Inner Join
```
Comment (sanderson) [inline stdlib/join/join.flux:43]:
```suggestion
// ### Perform an inner join
```

File: stdlib/join/join.flux [hunk 3]
```diff
@@ -1,15 +1,194 @@
...
+// join.tables(
+//     method: "inner",
+//     left: ints,
+//     right: strings,
+//     on: (l, r) => l._time == r._time,
+//     as: (l, r) => ({l with label: r._value}),
+// > )
+// ```
+//
+// ### Perform a Left Outer Join
```
Comment (sanderson) [inline stdlib/join/join.flux:60]:
```suggestion
// ### Perform a left outer join
```

File: stdlib/join/join.flux [hunk 4]
```diff
@@ -1,15 +1,194 @@
...
+// the output record is constructed in the `as` function.
+//
+// Because of how flux handles outer joins (see the `Outer Joins` section in the `join` [package
+// documentation](<insert-link-here>)), it's possible for either `l` or `r` to be a default record,
+// meaning any value in a non-group-key column could possibly be null.
+//
+// In the case of a left outer join, `l` is guaranteed to not be a default record. So, if we want to
+// ensure that the output record has non-null values for any columns that aren't part of the group
+// key, we should pull those values from `l`. By using a value from `r` that is not part of the
+// group key, you risk that value being null.
```
Comment (sanderson) [inline stdlib/join/join.flux:72]:
```suggestion
// In the case of a left outer join, `l` is guaranteed to not be a default record. To
// ensure that the output record has non-null values for any columns that aren't part of the group
// key, use values from `l`. Using a value from `r` that is not part of the
// risks that value being null.
```

File: stdlib/join/join.flux [hunk 5]
```diff
@@ -1,15 +1,194 @@
...
+//
+// In the example below, we construct the output record almost entirely from properties of `l`.
+// The exception comes when we define a proprty called `v_right` which gets its value from `r._value`.
+// In this case, we should understand and expect that `v_right` will sometimes be empty.
```
Comment (sanderson) [inline stdlib/join/join.flux:76]:
```suggestion
// The example below constructs the output record almost entirely from properties of `l`.
// The only exception is the `v_right` column which gets its value from `r._value`.
// In this case, understand and expect that `v_right` will sometimes be null.
```

File: stdlib/join/join.flux [hunk 6]
```diff
@@ -1,15 +1,194 @@
...
+// join.tables(
+//     method: "left",
+//     left: left,
+//     right: right,
+//     on: (l, r) => l.label == r.id and l._time == r._time,
+//     as: (l, r) => ({_time: l._time, label: l.label, v_left: l._value, v_right: r._value}),
+// > )
+// ```
+//
+// ### Perform a Right Outer Join
```
Comment (sanderson) [inline stdlib/join/join.flux:108]:
```suggestion
// ### Perform a right outer join
```

File: stdlib/join/join.flux [hunk 7]
```diff
@@ -1,15 +1,194 @@
...
+// join.tables(
+//     method: "right",
+//     left: left,
+//     right: right,
+//     on: (l, r) => l.label == r.id and l._time == r._time,
+//     as: (l, r) => ({_time: r._time, label: r.id, v_left: l._value, v_right: r._value}),
+// > )
+// ```
+//
+// ### Perform a Full Outer Join
```
Comment (sanderson) [inline stdlib/join/join.flux:145]:
```suggestion
// ### Perform a full outer join
```

File: stdlib/join/join.flux [hunk 8]
```diff
@@ -22,13 +201,21 @@ builtin tables : (
...
 //
 // ## Parameters
-// - left:
-// - right:
-// - as:
-// - method:
+// - left: Left input stream. Default is piped-forward data (<-).
+// - right: Right input stream.
+// - as: Function that takes a left and a right record (`l` and `r` respectively), and returns a record.
+//   The returned record will be included in the final output.
```
Comment (sanderson) [inline stdlib/join/join.flux:212]:
```suggestion
//   The returned record is included in the final output.
```

File: stdlib/join/join.flux [hunk 9]
```diff
@@ -54,13 +251,23 @@ inner = (left=<-, right, on, as) =>
...
-// - on:
-// - as:
+// - left: Left input stream. Default is piped-forward data (<-).
+// - right: Right input stream.
+// - on: Function that takes a left and right record (`l`, and `r` respectively), and returns a bool.
+//   The body of the function must be single boolean expression, consisting of one
+//   or more equality comparisons between a property of `l` and a property of `r`,
+//   each chained together by the `and` operator.
+// - as: Function that takes a left and a right record (`l` and `r` respectively), and returns a record.
+//   The returned record will be included in the final output.
```
Comment (sanderson) [inline stdlib/join/join.flux:266]:
```suggestion
//   The returned record is included in the final output.
```

File: stdlib/join/join.flux [hunk 10]
```diff
@@ -70,13 +277,23 @@ full = (left=<-, right, on, as) =>
...
-// - left:
-// - right:
-// - on:
-// - as:
+// - left: Left input stream. Default is piped-forward data (<-).
+// - right: Right input stream.
+// - on: Function that takes a left and right record (`l`, and `r` respectively), and returns a bool.
+//   The body of the function must be single boolean expression, consisting of one
+//   or more equality comparisons between a property of `l` and a property of `r`,
+//   each chained together by the `and` operator.
```
Comment (sanderson) [inline stdlib/join/join.flux:290]:
```suggestion
// - on: Function that takes a left and right record (`l`, and `r` respectively), and returns a boolean.
//
//   The body of the function must be a single boolean expression, consisting of one
//   or more equality comparisons between a property of `l` and a property of `r`,
//   each chained together by the `and` operator.
```

File: stdlib/join/join.flux [hunk 11]
```diff
@@ -86,13 +303,23 @@ left = (left=<-, right, on, as) =>
...
```
Comment (sanderson) [inline stdlib/join/join.flux:316]:
```suggestion
// - on: Function that takes a left and right record (`l`, and `r` respectively), and returns a boolean.
//
//   The body of the function must be a single boolean expression, consisting of one
//   or more equality comparisons between a property of `l` and a property of `r`,
//   each chained together by the `and` operator.
```

File: stdlib/join/join.flux [hunk 12]
```diff
@@ -86,13 +303,23 @@ left = (left=<-, right, on, as) =>
...
+// - as: Function that takes a left and a right record (`l` and `r` respectively), and returns a record.
+//   The returned record will be included in the final output.
```
Comment (sanderson) [inline stdlib/join/join.flux:318]:
```suggestion
//   The returned record is included in the final output.
```

**Review by sanderson — CHANGES_REQUESTED** [review]

File: stdlib/join/join.flux [hunk 13]
```diff
@@ -1,15 +1,194 @@
...
+// ## Metadata
+// introduced: NEXT
+// tags: transformations
 package join
 
 
-// tables is under active development and is not yet ready for public consumption.
+// tables joins together two input streams using a method, predicate, and a function that joins two records, one from each table.
+//
+// join.tables will only compare records that have the same group key. The tables in the output will have the same grouping as the inputs.
```
[the remark this answers is not in the document]
Comment (scbrickley) [reply stdlib/join/join.flux:27]: I'm not sure I follow. Could you elaborate a bit?

File: stdlib/universe/universe.flux [hunk 14]
```diff
@@ -1116,6 +1116,7 @@ builtin integral : (
 //
 // ## Metadata
 // introduced: 0.7.0
+// deprecated: NEXT
```
[the remark this answers is not in the document]
Comment (scbrickley) [reply stdlib/universe/universe.flux:1119]: So add the link in the `deprecated` field? As in `deprecated: NEXT, use [join.tables](url) instead`? Or somewhere else?

File: stdlib/join/join.flux [hunk 15]
```diff
@@ -1,15 +1,194 @@
...
+// Full outer joins make the guarantee for both input streams.
+//
+// Records in the specified table(s) that do not have a matching record in the other
+// input stream will instead be paired with a default record when running the function
+// defined in the `as` parameter.
+//
+// A default record has the same columns as the records in the corresponding input
+// table, but only the group key columns are populated with a value. All other
+// columns will have a null value.
+//
```
[the remark this answers is not in the document]
Comment (scbrickley) [reply stdlib/join/join.flux:18]: How's this?

```
// ## Inner joins
//
// Inner joins will drop any records that don't have a match in the other input stream.
// There is no need to account for default or unmatched records when performing an inner join.
```

[the remark this answers is not in the document]
Comment (sanderson) [reply stdlib/join/join.flux:18]: Sounds great. Only one minor change – update "will drop" to just "drop".

File: stdlib/universe/universe.flux (hunk 14 above)
[the remark this answers is not in the document]
Comment (sanderson) [reply stdlib/universe/universe.flux:1119]: No, add it to the function description. Something like this:

```
**Deprecated**: `join()` is deprecated in favor of [`join.inner()`](https://docs.influxdata.com/flux/v0.x/stdlib/join/inner/).
The [`join` package](https://docs.influxdata.com/flux/v0.x/stdlib/join/) provides support multiple join types.
```

File: stdlib/join/join.flux [hunk 16]
```diff
@@ -54,13 +317,77 @@ inner = (left=<-, right, on, as) =>
...
 // ## Parameters
-// - left:
-// - right:
-// - on:
-// - as:
+// - left: Left input stream. Default is piped-forward data (<-).
+// - right: Right input stream.
+// - on: Function that takes a left and right record (`l`, and `r` respectively), and returns a boolean.
+//
+//   The body of the function must be single boolean expression, consisting of one
```
Comment (sanderson) [inline stdlib/join/join.flux:329]:
```suggestion
//   The body of the function must be a single boolean expression, consisting of one
```

File: stdlib/experimental/experimental.flux [hunk 17]
```diff
@@ -316,6 +316,10 @@ builtin to : (
 
 // join joins two streams of tables on the **group key and `_time` column**.
 //
+// **Deprecated**: `experimental.join()` is deprecated in favor of [`join.inner()`](https://docs.influxdata.com/flux/v0.x/stdlib/join/inner/).
```
Comment (sanderson) [inline stdlib/experimental/experimental.flux:319]:
```suggestion
// **Deprecated**: `experimental.join()` is deprecated in favor of [`join.time()`](https://docs.influxdata.com/flux/v0.x/stdlib/join/time/).
```

**Review by sanderson — APPROVED** [review]
Just some last minute typos and then it's gtg 👍🏻

**Review by nathanielc — APPROVED** [review]

**Pull request merged**